sbi/diagnostics/tarp.py

```diff
@@ -0,0 +1,314 @@
+                 ``(num_samples, num_sims, num_dims)``. These are obtained by
+                 sampling a trained posterior `num_samples` times. Multiple
+                 (posterior) samples for one observation are encouraged.
+        theta: The true parameter value theta. Theta is expected to
+                 have shape ``(num_sims, num_dims)``.
+        references: the reference points to use for the coverage regions, with
+                shape ``(1, num_sims, num_dims)``, or ``None``.
+                If ``None``, then reference points are chosen randomly from
+                the unit hypercube over the parameter space given by theta.
+                In other words, reference samples are drawn from the
+                following ``Uniform(low=theta.min(dim=-1),high=theta.max(dim=-1))``.
+        distance: the distance metric to use when computing the distance.
+                Should be a callable function that accepts two tensors and
+                computes the distance between them, e.g. given two tensors
+                of shape ``(batch, 3)`` and ``(batch,3)``, this function should
+                return ``(batch,1)`` distance values.
+                Possible values: ``sbi.utils.metrics.l1`` or
+                ``sbi.utils.metrics.l2``. ``l2`` is the default.
+        num_bins: number of bins to use for the credibility values.
+                If ``None``, then ``num_sims // 10`` bins are used.
+        do_norm : whether to normalize parameters before coverage test
+                (Default = True)
+
+    Returns:
+        ecp: Expected coverage probability (``ecp``)
+        alpha: credibility values
+
+    """
+    # TARP assumes that the predicted thetas are sampled from the "true"
+    # PDF num_samples times
+    theta = theta.detach() if len(theta.shape) != 2 else theta.detach().unsqueeze(0)
+    samples = samples.detach()
+
+    assert (
+        theta.shape[-2:] == samples.shape[-2:]
+    ), f"shapes of theta {theta.shape[-2:]} and samples {samples.shape[-2:]} do not fit"
+
+    num_samples = samples.shape[0]  # samples per simulation
+    num_sims = samples.shape[-2]
+    num_dims = samples.shape[-1]
+
+    if num_bins is None:
+        num_bins = num_sims // 10
+
+    if theta.shape[-2] != num_sims:
+        raise ValueError("theta must have the same number of rows as samples")
+    if theta.shape[-1] != num_dims:
+        raise ValueError("theta must have the same number of columns as samples")
+
+    if do_norm:
+        lo = torch.min(theta, dim=-2, keepdims=True).values  # min along num_sims
+        hi = torch.max(theta, dim=-2, keepdims=True).values  # max along num_sims
+        samples = (samples - lo) / (hi - lo + 1e-10)
+        theta = (theta - lo) / (hi - lo + 1e-10)
+
+    references = _check_references(references, theta)
+    assert len(references.shape) == len(
+        samples.shape
+    ), f"references {references.shape} != samples {samples.shape}"
```

Comment:
Here I would just check whether references is a `Distritbution` or `None`. If the former, call `.sample((1,))` and assert that is matches `theta.shape[1:]`
If `None`, then construct the ref dist from the min max values (maybe in a separate function as it is now).

Reply:
I think this is a good idea. How strong are your feelings about this? If they are strong, I can put my hands to work on this. Currently, the `references` argument is not present in the signature of `run_tarp`. If we believe that this is handy, I can follow down this road tomorrow. But I also favor leaving this to a future release.

Reply:
I am fine with fixing this later.